src_c/math.c

```diff
@@ -1633,18 +1633,49 @@ vector_reflect_ip(pgVector *self, PyObject *normal)
 static double
 _vector_distance_helper(pgVector *self, PyObject *other)
 {
-    Py_ssize_t i;
-    double distance_squared, tmp;
+    Py_ssize_t i, dim = self->dim;
+    double distance_squared = 0;
+
+    if (pgVector_Check(other)) {
+        pgVector *otherv = (pgVector *)other;
+        double dx, dy;
+
+        if (dim != otherv->dim) {
+            PyErr_SetString(PyExc_TypeError, "Vectors must be the same size");
+            return -1;
+        }
 
-    distance_squared = 0;
-    for (i = 0; i < self->dim; ++i) {
-        tmp = PySequence_GetItem_AsDouble(other, i) - self->coords[i];
-        distance_squared += tmp * tmp;
+        dx = otherv->coords[0] - self->coords[0];
+        dy = otherv->coords[1] - self->coords[1];
+
+        distance_squared = dx * dx + dy * dy;
+
+        if (dim == 3) {
+            double dz;
+            dz = otherv->coords[2] - self->coords[2];
+            distance_squared += dz * dz;
+        }
     }
-    /* PySequence_GetItem_AsDouble can fail in which case it will set an Err */
-    if (PyErr_Occurred())
-        return -1;
+    else {
+        double tmp;
+        if (!(PyTuple_Check(other) || PyList_Check(other))) {
+            PyErr_SetString(PyExc_TypeError, "A sequence was expected");
+            return -1;
+        }
+        if (PySequence_Fast_GET_SIZE(other) != dim) {
+            PyErr_SetString(PyExc_TypeError,
+                            "Vector and sequence must be the same size");
+            return -1;
+        }
 
+        for (i = 0; i < self->dim; ++i) {
```

Comment:
Just a minor nitpick, this can be just `dim` so that it is consistent with other code in this function